### PR TITLE
[Impeller] Optionally support SamplerAddressMode::kDecal on the OpenGLES backend

### PR DESCRIPTION
**Pull request by ColdPaleLight**
fix https://github.com/flutter/flutter/issues/129358

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

### CODE REVIEW

File: impeller/renderer/backend/gles/sampler_gles.cc [hunk 1]
```diff
@@ -53,7 +53,8 @@ static GLint ToParam(MinMagFilter minmag_filter,
   FML_UNREACHABLE();
 }
 
-static GLint ToAddressMode(SamplerAddressMode mode) {
+static GLint ToAddressMode(SamplerAddressMode mode,
```
Comment (jonahwilliams) [inline impeller/renderer/backend/gles/sampler_gles.cc:56]: nit: I might just pass in a bool for whether or not decal is supported.

**Review by jonahwilliams — COMMENTED** [review]
Nice!

There are a few places where we ifdef in the shader, such as

https://github.com/flutter/engine/blob/a18c52768bbc8c779604a186dc6f9d14e73cee14/impeller/entity/shaders/blending/porter_duff_blend.frag#L32

I think we should be able to update the ifdef to check for the extension? or something like that.

File: impeller/renderer/backend/gles/test/capabilities_unittests.cc [hunk 2]
```diff
@@ -34,5 +34,15 @@ TEST(CapabilitiesGLES, CanInitializeWithDefaults) {
             PixelFormat::kD24UnormS8Uint);
 }
 
+TEST(CapabilitiesGLES, SupportsDecalSamplerAddressMode) {
+  auto const extensions = std::vector<unsigned char*>{
+      (unsigned char*)"GL_KHR_debug",                 //
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/test/capabilities_unittests.cc:39]: Nit: `reinterpret_cast`

File: impeller/renderer/backend/gles/sampler_gles.cc (hunk 1 above)
Comment (chinmaygarde) [reply impeller/renderer/backend/gles/sampler_gles.cc:56]: Just a drive by comment; Is it just me or does anyone else find "decal" mode to be a non-standard and/or overly confusing term? How about just `ClampToBorderColor`? Thats what the various APIs call it (`GL_CLAMP_TO_BORDER_EXT`, `MTLSamplerAddressModeClampToBorderColor`, `VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_BORDER`).

Perhaps later, we can rename this enum value for consistency.

File: impeller/renderer/backend/gles/sampler_gles.cc [hunk 3]
```diff
@@ -96,10 +100,12 @@ bool SamplerGLES::ConfigureBoundTexture(const TextureGLES& texture,
                    ToParam(desc.min_filter, mip_filter));
   gl.TexParameteri(target.value(), GL_TEXTURE_MAG_FILTER,
                    ToParam(desc.mag_filter));
-  gl.TexParameteri(target.value(), GL_TEXTURE_WRAP_S,
-                   ToAddressMode(desc.width_address_mode));
-  gl.TexParameteri(target.value(), GL_TEXTURE_WRAP_T,
-                   ToAddressMode(desc.height_address_mode));
+  gl.TexParameteri(
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/sampler_gles.cc:103]: How do we set the border color? `TEXTURE_BORDER_COLOR_EXT`

File: impeller/renderer/backend/gles/sampler_gles.cc (hunk 1 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/sampler_gles.cc:56]: There is a cost for non-consistency with the underlying graphics API, but there is also a value in matching the framework level enumeration which uses decal as a terminology. I don't feel as strongly about decal as I do about "FilterQuality", which is entirely opaque - at least decal has some semantic meaning

File: impeller/renderer/backend/gles/sampler_gles.cc (hunk 3 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/sampler_gles.cc:103]: Maybe this defaults to transparent black so we don't need to worry about it?

Comment (jonahwilliams) [reply impeller/renderer/backend/gles/sampler_gles.cc:103]: according to https://registry.khronos.org/OpenGL-Refpages/es3/html/glTexParameter.xhtml this value defaults to transparent black so we don't need to set anything.

Comment (chinmaygarde) [reply impeller/renderer/backend/gles/sampler_gles.cc:103]: I guess. I took a look at Vulkan and [setting the border color is an extension](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_custom_border_color.html). But setting it to `VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_BORDER` is not. I guess this is fine. I don't think we set the color on Metal either. Just really odd.

Comment (jonahwilliams) [reply impeller/renderer/backend/gles/sampler_gles.cc:103]: Flutter does not expose an API for setting the border color.

Comment (chinmaygarde) [reply impeller/renderer/backend/gles/sampler_gles.cc:103]: Nevermind. Its an extension that is core in 1.1. And we do set it to [transparent black](https://github.com/flutter/engine/blob/a18c52768bbc8c779604a186dc6f9d14e73cee14/impeller/renderer/backend/vulkan/sampler_library_vk.cc#L46). Let's be explicit and do the same here as well please?

Comment (chinmaygarde) [reply impeller/renderer/backend/gles/sampler_gles.cc:103]: Its is also dealers choice on Metal at the moment. Let's be explicit and set borderColor [here too](https://github.com/flutter/engine/blob/a18c52768bbc8c779604a186dc6f9d14e73cee14/impeller/renderer/backend/metal/sampler_library_mtl.mm#L25C17-L25C37). In fact, a default border color specification in `formats.h` would be good for consistency.

Comment (chinmaygarde) [reply impeller/renderer/backend/gles/sampler_gles.cc:103]: > Flutter does not expose an API for setting the border color.

Yeah, just didn't want to read the spec to see if the default was transparent black. Just being explicit seemed better. There isn't any bug though.

File: impeller/renderer/backend/gles/sampler_gles.cc (hunk 1 above)
Comment (ColdPaleLight) [reply impeller/renderer/backend/gles/sampler_gles.cc:56]: > nit: I might just pass in a bool for whether or not decal is supported.

Done

File: impeller/renderer/backend/gles/test/capabilities_unittests.cc (hunk 2 above)
Comment (ColdPaleLight) [reply impeller/renderer/backend/gles/test/capabilities_unittests.cc]: Done

File: impeller/renderer/backend/gles/sampler_gles.cc (hunk 3 above)
Comment (ColdPaleLight) [reply impeller/renderer/backend/gles/sampler_gles.cc:103]: Issue filed: https://github.com/flutter/flutter/issues/136153

**Comment by ColdPaleLight** [comment]
> Nice!
> 
> There are a few places where we ifdef in the shader, such as
> 
> https://github.com/flutter/engine/blob/a18c52768bbc8c779604a186dc6f9d14e73cee14/impeller/entity/shaders/blending/porter_duff_blend.frag#L32
> 
> I think we should be able to update the ifdef to check for the extension? or something like that.

I removed the macro and added a variable in the uniform block to determine if the gpu api supports clip to border. (I originally planned to use different pipelines that support decal and those that do not, but I found that this would double the number of advanced blend pipelines.)

**Comment by jonahwilliams** [comment]
Rather than use a uniform, can we use a combination of defines - if opengles and at least one of the decal extensions not defined? Each extension should have a corresponding define already

**Comment by ColdPaleLight** [comment]
> Rather than use a uniform, can we use a combination of defines - if opengles and at least one of the decal extensions not defined? Each extension should have a corresponding define already

I guess this won't work because shaderc will do a preprocessing before compiling glsl, so the macros will be handled at that stage. Therefore, glsl generated by spirv cross no longer contains statements like "#ifdef GL_EXT_texture_border_clamp"

**Comment by jonahwilliams** [comment]
that is extremely unfortunate

**Comment by jonahwilliams** [comment]
Well, we can clean it up later then. There is probably another way to do it by way of tinkering with SPIRV cross.

File: impeller/entity/shaders/blending/advanced_blend.glsl [hunk 4]
```diff
@@ -24,12 +25,11 @@ in vec2 v_src_texture_coords;
 out f16vec4 frag_color;
 
 f16vec4 Sample(f16sampler2D texture_sampler, vec2 texture_coords) {
-// gles 2.0 is the only backend without native decal support.
-#ifdef IMPELLER_TARGET_OPENGLES
-  return IPSampleDecal(texture_sampler, texture_coords);
-#else
-  return texture(texture_sampler, texture_coords);
-#endif
+  if (blend_info.supports_decal_sampler_address_mode > 0.0) {
```
Comment (jonahwilliams) [inline impeller/entity/shaders/blending/advanced_blend.glsl:28]: These can still technically be ifdefs since vulkan/metal always support this.

Comment (ColdPaleLight) [reply impeller/entity/shaders/blending/advanced_blend.glsl:29]: Got it, added them back.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**